sql-plugin/src/main/scala/com/nvidia/spark/rapids/AbstractGpuJoinIterator.scala

```diff
@@ -179,6 +179,13 @@ abstract class SplittableJoinIterator(
   protected def createGatherer(cb: ColumnarBatch, numJoinRows: Option[Long]): Option[JoinGatherer]
 
   override def hasNextStreamBatch: Boolean = {
+    // make built batch spillable
+    if (isInitialJoin) {
```

Comment:
This is not be specific to the initial join, as this problem can occur on every stream iterator query.  This is really  all about `stream.hasNext`.  Since we typically manifest our batches (and perform shuffle pulls) during `hasNext`, that's when we want to release the semaphore.  Therefore I would expect it to look more like this:
```
if (pendingSplits.nonEmpty) {
  // there's at least one cached stream-side batch ready to go
  true
} else {
  // Querying the stream side may involve network or other non-GPU computation,
  // so ensure our memory buffers are spillable and release the GPU to allow other
  // threads to use the GPU in the interim. The stream iterator will re-grab the GPU
  // semaphore when it's ready to produce a batch.
  builtBatch.allowSpilling()
  GpuSemaphore.releaseIfNecessary(TaskContext.get())
  stream.hasNext || isInitialJoin
}
```

Reply:
Thanks @jlowe! I will make the suggested change.

Reply:
I put up this change, but I'm not sure I fully understand why just doing it on the initialJoin isn't sufficient?   It doesn't look like builtBatch changes after we construct this iterator.

Reply:
This isn't so much about builtBatch as it is releasing the semaphore.  We need to call `allowSpilling` before the first iteration (or the caller of the iterator needs to do it before passing the lazy spillable to the iterator), so that's why we're doing that here.  After the first iteration, `allowSpilling` should be getting called in `nextCbFromGatherer` as it produces the next output batch.  `allowSpilling` is idempotent, so shouldn't matter that we're doing it here even after the first iteration.

However releasing the semaphore seems key to getting the performance win.  IIUC the performance problem was holding the lock while trying to fetch the next batch from the stream iterator and that fetch may involve network shuffle fetch or other non-GPU activity.  To turn the question around a bit: why would we only want to release the GPU semaphore when fetching the first stream batch instead of consistently doing it before fetching every stream batch?

Reply:
My understanding (which may be flawed), is that the issue is with a join where the first half needed the GPU and produced builtBatch, but the second half does not need the GPU.  The case we were seeing was that the GPU semaphore was being held the entire time.   It's not clear to me why we would need to release the semaphore more than once at the beginning of processing the stream side?

Reply:
In general, pulling batches from an input iterator may involve performing non-GPU activities.  This is not a problem for "simple" exec nodes where they immediately process an input batch to produce an output batch.  They're not caching any data locally and don't have code to acquire or release the semaphore.  However if an exec node is more complicated, where it can cache data on the GPU in-between fetching from input iterators, then we need to consider carefully what happens with the semaphore.

Fetching from the build side is not so much a problem because it fetches exactly one batch (due to a GpuCoalesceExec being placed before it in the plan on the build side which guarantees all data will be coalesced into a single batch).  GpuCoalesceExec in that case is responsible for releasing the semaphore in-between fetches as it caches the data, and if you look at that code you'll see that logic.

So that's the build side.  However after pulling from the build side we now have a build-side batch sitting in GPU memory and go to fetch from the stream side.  Fetching from the stream side may involve (and often does in the case of join) fetching from a shuffle iterator.  We don't want to hold the GPU semaphore when fetching from the stream side and it involves doing network, RowToColumnExec (i.e.: executing part of the query on the CPU), or other non-GPU activity.  However we can't just release the semaphore anytime we feel like it, because we have this build-side batch sitting in GPU memory and it's not yet spillable.  That's why we call `allowSpilling` on it before releasing the semaphore, because that allows the data we want to cache to be released on-demand if needed.

> It's not clear to me why we would need to release the semaphore more than once at the beginning of processing the stream side?

That's because the stream side, unlike the build side, is not guaranteed to arrive in a single batch.  It could be many batches and therefore many iterations.  The same concerns we have on the first batch can happen on every batch.  If we only release the semaphore on the first iteration, we only get the benefit on the first stream-side batch.  That's great if there's only one batch, but if there's more than one batch then we will continue to hold the semaphore while potentially processing network traffic or CPU-produced batches on the stream side for subsequent batches.

Reply:
Thanks for the explanations @jlowe.  I was looking at q23b, which appears to effectively only have one batch per task, so I wasn't seeing the bigger picture.